spec.bs

```diff
@@ -461,6 +473,146 @@ An <dfn>internal observer</dfn> is a [=struct=] with the following [=struct/item
   [[#promise-returning-operators]] that make use of this, for example.</p>
 </div>
 
+<div algorithm>
+  To <dfn for=Observable>convert to an Observable</dfn> given an {{any}} |value|, run these steps:
+
+  Note: We split this algorithm out from the Web IDL {{Observable/from()}} method, so that
+  spec prose can <a for=Observable lt="convert to an observable">convert</a> an {{Observable}}
+  without going through the Web IDL bindings.
+
+  Note: The resolution of value to its descrete types happens before
+  [=Observable/subscribe callback=] is called. This means mutations of values, such as adding
+  the iterable protocols to the object, will not take affect between the creation of the returned
+  observable, and when it is subscribed to.
+
+    1. If [$Type$](|value|) is not [=Object=], [=exception/throw=] a {{TypeError}}.
+
+       Note: This prevents primitive types from being coerced into iterables (e.g., String).
+
+       Issue: See if this is even the behavior we want. See <a
+       href=https://github.com/WICG/observable/issues/125>WICG/observable#125</a>
+
+    1. <i id=from-observable-conversion><b>From Observable</b></i>: If |value|'s [=specific type=]
+       is an {{Observable}}, then return |value|.
+
+    1. <i id=from-async-iterable-conversion><b>From async iterator</b></i>: Let
+       |asyncIteratorMethodRecord| be [=?=] [$GetMethod$](|value|, {{%Symbol.asyncIterator%}}).
+
+       Note: We have to use [$GetMethod$] directly instead of [$GetIterator$], because
+       [$GetIterator$] makes it impossible to distinguish between (a) the case where |value| simply
+       doesn't implement any iterable protocol, and (b) the case where |value| implements one of the
+       iterator protocols, and its method getter throws an exception.
+
+       Note: This step is web-observable, and re-throws any errors that the {{%Symbol.iterator%}}
+       method *getter* might have thrown.
+
+    1. If |asyncIteratorMethodRecord|'s \[[Value]] is undefined or null, then jump to the step
+       labeled <a href=#from-iterable-conversion>From iterable</a>.
+
+    1. If [$IsCallable$](|asyncIteratorMethodRecord|'s \[[Value]]) is false, then
+       [=exception/throw=] a {{TypeError}}.
+
+       Otherwise, let |nextAlgorithm| be the following steps, given |subscriber| and |iterator|:
```

Comment:
I've cleaned up the async iterator conversion logic a good bit, but I think one thing we're still not doing is checking for DONE-ness. I think at some point we need to be checking if the iterable is DONE, i.e., a line like https://streams.spec.whatwg.org/#readable-stream-from-iterable:~:text=Let%20done%20be%20%3F%20IteratorComplete(iterResult).

Reply:
I missed it, but I guess we are doing that in the first step of `nextAlgorithm`. But I think the handling of this is wrong anyways. See the other comment below about moving the done check.